### PR TITLE
Autoderef when suggesting to call `(self.field)`

### LINKED ISSUE #32128: Misleading note for accessing closures within structs.
Hi, I encountered an issue the other day wherein I was attempting to call a closure stored within a struct. 

The compiler gives a misleading error which doesn't give the user an idea on how to fix the problem.
I've prepared an example which reproduces the misleading text:

```
struct Example {
    example: Box<Fn(i32) -> i32>
}

fn main() {
    let demo = Example {
        example: Box::new(|x| {
            x + 1
        })
    };
    
    demo.example(1);
    // (demo.example)(1);
}
```

The error this yields from the compiler is as follows:

```
structs.rs:12:10: 12:20 error: no method named `example` found for type `Example` in the current scope
structs.rs:12     demo.example(1);
                       ^~~~~~~~~~
structs.rs:12:10: 12:20 note: did you mean to write `demo.example`?
structs.rs:12     demo.example(1);
                       ^~~~~~~~~~
error: aborting due to previous error
```

Note that in the source code I've provided, underneath the line which yields this error, I have provided a comment demonstrating the correct way to access and call the example member of the structure, however the compiler's example is identical to the erroneous code and thus doesn't help the user in working out how to properly call the member.

Would it be at all possible to improve this error message?

I've confirmed this is reproducible on `rustc 1.7.0 (a5d1e7a59 2016-02-29)`.

### PR DESCRIPTION
**Pull request by jonas-schievink**
Fixes #32128

### CODE REVIEW

File: src/librustc_typeck/check/method/suggest.rs [hunk 1]
```diff
@@ -50,23 +51,40 @@ fn is_fn_ty<'a, 'tcx>(ty: &Ty<'tcx>, fcx: &FnCtxt<'a, 'tcx>, span: Span) -> bool
...
+
+                        if selcx.evaluate_obligation(&obligation) {
+                            Some(true)
+                        } else {
+                            None
+                        }
+                    })
+                });
+
+                match opt_is_fn {
```
Comment (nagisa) [inline src/librustc_typeck/check/method/suggest.rs]: `opt_is_fn.unwrap_or(false)`?

Comment (jonas-schievink) [inline src/librustc_typeck/check/method/suggest.rs]: Gah, of course. Fixed.

File: src/librustc_typeck/check/method/suggest.rs [hunk 2]
```diff
@@ -50,23 +51,37 @@ fn is_fn_ty<'a, 'tcx>(ty: &Ty<'tcx>, fcx: &FnCtxt<'a, 'tcx>, span: Span) -> bool
...
+                                            cx.mk_substs(fn_once_substs));
+                        let poly_trait_ref = trait_ref.to_poly_trait_ref();
+                        let obligation = Obligation::misc(span,
+                                                          fcx.body_id,
+                                                          poly_trait_ref
+                                                             .to_predicate());
+                        let mut selcx = SelectionContext::new(infcx);
+
+                        if selcx.evaluate_obligation(&obligation) {
+                            Some(true)
```
Comment (nagisa) [inline src/librustc_typeck/check/method/suggest.rs]: Nit: I can’t find `Some(false)` ever getting emitted. Couldn’t you just emit `Some(())` and then interpret the discriminant as boolean with `.is_some()` instead?

**Comment by nagisa** [comment]
This seems good to me. r=me with nit fixed or well objected to.

File: src/librustc_typeck/check/method/suggest.rs (hunk 2 above)
Comment (jonas-schievink) [inline src/librustc_typeck/check/method/suggest.rs]: Yes, sure. I just got caught up in a mess of `bool`s I think :-)

This should be better.

**Pull request merged**